architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+#### Resource Types
+
+The resource type registration APIs add two new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+
+These new resource types define the full CRUDL set of APIs. These APIs are optimized for operations that need to modify or deeply understand resource types. 
+
+`resourceTypes` is our first nested resource type in UCP/Radius. This is a good usecase for a nested resource types because:
+
+- The `resourceTypes` body will be be large due to the inclusion of an OpenAPI schema. 
+- Users need granular registration and unregistration of individual resource types. (See scenario 2 above).
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
+
+The rationale for this is to support registration/unregistration per-tenant of Radius rather than per-installation.
+
+#### Resource Type Queries
+
+The proposal also defines two new **non-resource** APIs:
+
+- `GET /planes/radius/{plane}/providers/`: list all resource providers.
+- `GET /planes/radius/{plane}/providers/{provider}` shows a specific resource provider.
+
+These APIs are optimized for queries and are consistent with similar APIs in ARM. To understand why this is important, remember that the OpenAPI schema for a resource type can be quite large. Some functionality in Radius will need to query resource types frequently, consider the need for the `rad recipe register` to validate a resource type. Building a separate API for queries allows us to optimize the behavior.
+
+These APIs can also be implemented consistently across all planes in the future. For example Azure doesn't support user-defined types, but it does support looking up the set of resource types.
+
+Lastly, there will still be some system-defined types. For example `System.Resources/resourceProviders` is a system defined type. The query API can return these, but the `System.Resources/resourceProviders` cannot. Imagine for a minute what would happen if users could remove the capability to specify a type.
+
+**Open Question:** Is *query* the right term for this? The term will show up in docs and in the TypeSpec + generated client + our code. Let's pick the term that's least confusing to us. I'm open to suggestions but I'll just roll with *query* unless there's a better idea.
+
+#### Locations
+
+UCP is a multi-location control-plane like the control-plane that inspired it (ARM). We don't use UCP/Radius in any multi-location scenarios right now but it's important we don't design APIs that lock us into a single-region design. ARM has a mature multi-location design but doesn't expose to users the ability to register types. However we can learn a lot from the scenarios supported by ARM for rolling out new API versions and new types.
```

Comment:
Can we talk about what `multi-location` means? Are we talking about actual cloud locations/regions? Or are we talking something like multi-cluster? And why are the resource providers deployed in a single location? If we have a multi-location UCP and register a resource type using UCP, why is that resource type not available to all the locations? An example could be nice here.

Reply:
And you gave the example below 🤣

Reply:
But why are we limiting the resource types to a single location?

Reply:
I would also like to learn more about UCP being multi-location.

Reply:
The concept of `location` in UCP/ARM answers the question:

> What instance of the control-plane is responsible for this resource?

Remember that `location` does not constrain what users can do. Users **can** mix and match the location of the control plane with the location of the resources they create. 

> Or are we talking something like multi-cluster?

Not exactly. Each location is a "logical instance" of Radius that's working together with others. Like a player on a soccer field, they move independently but they are all playing the same game.

For naive users they will use a single location for everything 👍.

---

`location` provides a lot of value when you start to optimize around it. Being multi-location allows users to deploy UCP/Radius next to the resources they want it to manage. This limits the amount of cross-zonal network traffic required and removes single points of failure.

Imagine you deployed Radius in `westus3`, but you're deploying applications in `westus3`, `canadaeast` and `westeurope`. You *could* do this and it would work.  However if `westus3` has an outage then you can't manage **any** of your applications. 

Now imagine you create a multi-location deployment of Radius. You create a location for each logical location where you want to run applications (`westus3`, `canadaeast` and `westeurope`). Now if `westus3` has an outage you can still manage applications in `canadaeast` and `westeurope`.

Reply:
I added more definitions and information.